### ISSUE TITLE
When AdBlock Plus is used, suggest user to disable it for Piwik

### ISSUE THREAD

**Issue by mattab** [piwik/piwik#5094]
Adblock plus creates issues when used for Piwik app. AdblockPlus blocks some javascript from executing because of their filenames and/or path. It catches false positive in the Piwik user interface.

This was reported many times over, [for example here](http://forum.piwik.org/read.php?2,114755).

Proposal 
* Detect adblock plus is used, and show a warning/notification on the screen asking to disable it
* Contact Adblock plus team, and kindly ask them to fix their regex for piwik

**Comment by tsteur** [comment]
:+1:  for this. Many will open Piwik and even demo.piwik.org and just think that it doesn't work

**Comment by hpvd** [comment]
one should do this detection **really carefully**
There are lot's of methods out there which work with adding blacklisted words e.g. "advertise" to the website code the see if this is blocked.
Doing this here would be like honey and we will have problems with many many other other tools/extension...

**Comment by hpvd** [comment]
I would suggest to understand **why piwik-backend is blocked**
- for website Fronend with piwik script insited - it's clear
- but why are ther know problems with Piwik's backend?

**Comment by hpvd** [comment]
so it should be possible to completly disable this function an making EVERY SINGLE part of this code invisible on the website (for those who would like to play with honey when there are bees outside)

**Comment by mattab** [comment]
Also experienced in forums: http://forum.piwik.org/read.php?2,129115

> Ublock Origin is now blocking the content within Piwik admin area.
> Add your admin panel url ot the white list of Ublock Origin, restart the browser and it works again. 
> Maybe it worth to add it to the FAQ as Ublock Origin is quite popular now.

**Comment by mattab** [comment]
There are tricks to detect adblocker is running eg. http://stackoverflow.com/a/24111206

But  we want to detect whether the adblocker is currently blocking some of the Piwik JS content, as many ad blockers will correctly let Piwik dashboard render, so we shouldn't notify users in that case.

**Comment by ThaDafinser** [comment]
There are 100 ways javascript or Piwik can be broken. So i would suggest a more general solution.

What about saving a list of the .js files which should get loaded in the HTML main file.
Then after the onload event is fired check if all are loaded.

When something is wrong, then display the user some possible solutions: `Piwik cannot work, because some Scripts werde blocked. Do you have enabled AdBlock? Do you have enable another blocking addon?`

**Comment by tsteur** [comment]
When development mode is disabled, we have only one or two files to load but I like that idea which I had not thought about before. Eg we could check something like `if !$('#content:visible')`  or `if !$('#content').text()`. Not sure if it works though. Need to check how they block the content exactly

**Comment by tsteur** [comment]
I checked Dashboard and Admin area with  Ublock Origin extension in Chrome and Firefox, with development mode enabled and disabled but couldn't reproduce. Can anyone reproduce it and let me know if there are any configurations to be done?

**Comment by tsteur** [comment]
FYI: The following terms are blocked eg by [easyprivacy list](https://easylist-downloads.adblockplus.org/easyprivacy.txt): 

```
/piwik-$domain=~piwik.org
/piwik.$script,domain=~piwik.org
/piwik.php
/piwik/js/*$domain=~piwik.org
/piwik1.
/piwik2.js
/piwik_
/piwikapi.js
/piwikC_
/piwikTracker.
```

So using eg this URL results in no js file being loaded:  `http://apache.piwik/piwik_1/`

The requests actually start but never finish, not sure how we can detect this in general. Also it always depends on the Piwik URL etc which Piwik files will be loaded and which not. Eg when using `http://apache.piwik` as URL all files are loaded unless one enables development mode. Then eg `piwik-api.js` won't be loaded but also 2 other files. 

![adblock](https://cloud.githubusercontent.com/assets/273120/10371375/580b104a-6de3-11e5-8adc-2a3a5066e77e.png)

**Comment by tsteur** [comment]
I worked on a detection but it's not trivial as maybe all JavaScript files are blocked and we cannot rely on anything. 

How it works:
* To have it work in development mode enabled and disabled I set `var hasBlockedContent = false;` in `piwik-api.js`
* Add the end of the page we check whether `hasBlockedContent` was set to `false` and if not, there might be an ad blocker used. It could be as well a 404 or some other problems that are not related to ad block
* To increase the chances that there is an ad block active we check whether an element having the id `bottomAd` was set to `display:none` or has `height:0`. If this is the case we presume there is an ad block active and show a red, bold message at the bottom. The message is at the bottom in case we detected it wrong to keep. One will still notice it at some point, especially if all content was blocked.

*Why do we not only do the `bottomAd` check?*
We would maybe detect an adblock that does not block any of our files, in this case we should not show the message as our detection of `bottomAd` doesn't have to be 100% right.

*Do we detect all ad blocker?*
Maybe, possibly not. I don't know if all Ad blocker work like "uBlock Origin" by blocking requests. Also we only show the warning if we detected the `bottomAd` block correctly. There might be ad blocker that do this differently. I think it's better to show the red warning message only when we are very sure about an ad blocker being active and not when there are maybe other reasons like an HTTP 404.

*How does the ad blocker block our content?*
See screenshot above but it depends on the browser. In Chrome an error "ERR_BLOCKED_BY_CLIENT" appears in the developer tools. I tried to detect such errors but seems not possible. Eg tried by using `window.onerror`, `script.onerror` and many others without success. Also did not find something about this.

*How does the warning look?*
It depends on which files are loaded as it might load some files in development mode and result can very. 

When all files are blocked it possibly looks like this
<img width="469" alt="screenshot at oct 082" src="https://cloud.githubusercontent.com/assets/273120/10373951/c478be96-6df0-11e5-9ceb-71114f879162.png">

If only some files are blocked it can look like this
<img width="653" alt="screenshot at oct 083" src="https://cloud.githubusercontent.com/assets/273120/10373976/db9c0b1e-6df0-11e5-80d5-c6060efd725a.png">
